File: scratch/dedup.py
import re
import json
import os
import ast

file_path = r"d:\new\cyber-chell\src\data\terms-data.ts"

with open(file_path, "r", encoding="utf-8") as f:
    content = f.read()

# Extract the array part
match = re.search(r"export const terms:\s*Term\[\]\s*=\s*\[(.*?)\];", content, re.DOTALL)
if not match:
    print("Could not find terms array")
    exit()

array_content = match.group(1)

# We can parse the objects using a simple state machine or regex since it's JS objects
# Alternatively, since the file seems to have JSON-like objects (mostly string keys or unquoted keys)
# Let's clean it up to be valid JSON
import js2py # wait, we might not have js2py. 
# Let's just use regex to extract objects
objects_raw = re.findall(r"\{.*?\}", array_content, re.DOTALL)

terms = []
for obj_str in objects_raw:
    # Extract keys and values
    # e.g., { id: 1, termEn: "Access Control", ... }
    obj = {}
    id_m = re.search(r'(?:"id"|id):\s*(\d+)', obj_str)
    en_m = re.search(r'(?:"termEn"|termEn):\s*"([^"]+)"', obj_str)
    ar_m = re.search(r'(?:"termAr"|termAr):\s*"([^"]+)"', obj_str)
    def_ar_m = re.search(r'(?:"definitionAr"|definitionAr):\s*"([^"]+)"', obj_str)
    def_en_m = re.search(r'(?:"definitionEn"|definitionEn):\s*"([^"]+)"', obj_str)
    ex_m = re.search(r'(?:"example"|example):\s*"([^"]+)"', obj_str)
    lvl_m = re.search(r'(?:"level"|level):\s*"([^"]+)"', obj_str)
    cat_m = re.search(r'(?:"categoryId"|categoryId):\s*(\d+)', obj_str)
    
    if id_m and en_m and ar_m:
        obj['id'] = int(id_m.group(1))
        obj['termEn'] = en_m.group(1)
        obj['termAr'] = ar_m.group(1)
        if def_ar_m: obj['definitionAr'] = def_ar_m.group(1)
        if def_en_m: obj['definitionEn'] = def_en_m.group(1)
        if ex_m: obj['example'] = ex_m.group(1)
        if lvl_m: obj['level'] = lvl_m.group(1)
        if cat_m: obj['categoryId'] = int(cat_m.group(1))
        
        # also store original string for replacement
        obj['_raw'] = obj_str
        terms.append(obj)

print(f"Total terms extracted: {len(terms)}")

# Deduplication logic
def normalize(text):
    text = text.strip().lower()
    text = re.sub(r'[^a-z0-9\s-]', '', text)
    text = re.sub(r'\s+', ' ', text)
    return text

unique_terms = {}
duplicates = []

for term in terms:
    norm_en = normalize(term['termEn'])
    
    if norm_en in unique_terms:
        duplicates.append((term, unique_terms[norm_en]))
    else:
        unique_terms[norm_en] = term

print(f"Duplicates found: {len(duplicates)}")

# Output summary to a json file
out_data = {
    "total_before": len(terms),
    "total_after": len(unique_terms),
    "duplicates_removed": len(duplicates),
    "removed_list": [{"removed": d[0]['termEn'], "kept": d[1]['termEn']} for d in duplicates]
}

with open("dedup_summary.json", "w", encoding="utf-8") as f:
    json.dump(out_data, f, ensure_ascii=False, indent=2)

print("Summary saved to dedup_summary.json")

File: scratch/dedup.js
const fs = require('fs');

const filePath = 'src/data/terms-data.ts';
let content = fs.readFileSync(filePath, 'utf-8');

// The file exports `terms` array. We can use eval to get the array, or extract it via regex.
// Since it's a TS file with `export const terms: Term[] = [...]`, we can extract the JSON-like array.
// But it has unquoted keys or single quotes. Let's write a small parser or use a safer approach.
// We can strip the export part, compile it using tsc or just eval it after some replacements.
// Since we want to maintain the formatting, doing regex might be better or we just recreate the TS file.
// The file seems to just be an array of objects.

// Let's use a regex approach to extract each object.
const objectRegex = /{\s*(?:id|\"id\"):\s*(\d+).*?}/gs;
const matches = [...content.matchAll(objectRegex)];

let terms = [];
for (const match of matches) {
    const objStr = match[0];
    
    // Extract fields
    const getFieldStr = (fieldName) => {
        const regex = new RegExp(`(?:${fieldName}|\\"${fieldName}\\")\\s*:\\s*(?:\\"([^\\"]+)\\"|\\'([^\\']+)\\')`);
        const m = objStr.match(regex);
        return m ? (m[1] || m[2]) : null;
    };
    
    const getFieldNum = (fieldName) => {
        const regex = new RegExp(`(?:${fieldName}|\\"${fieldName}\\")\\s*:\\s*(\\d+)`);
        const m = objStr.match(regex);
        return m ? parseInt(m[1]) : null;
    };

    terms.push({
        id: getFieldNum('id'),
        termEn: getFieldStr('termEn'),
        termAr: getFieldStr('termAr'),
        definitionAr: getFieldStr('definitionAr'),
        definitionEn: getFieldStr('definitionEn'),
        example: getFieldStr('example'),
        level: getFieldStr('level'),
        categoryId: getFieldNum('categoryId'),
        _raw: objStr
    });
}

console.log(`Extracted ${terms.length} terms`);

function normalize(text) {
    if (!text) return '';
    let t = text.trim().toLowerCase();
    // remove special chars if not meaningful
    t = t.replace(/[^a-z0-9\s-]/g, '');
    t = t.replace(/\s+/g, ' ');
    return t;
}

// Semantic similarities can be added if needed
const semanticMap = {
    "login": "sign in",
    "sign in": "sign in",
    "log in": "sign in"
};

let uniqueTerms = new Map();
let duplicates = [];
let cleanedTerms = [];

for (const term of terms) {
    let normEn = normalize(term.termEn);
    if (semanticMap[normEn]) {
        normEn = semanticMap[normEn];
    }
    
    if (uniqueTerms.has(normEn)) {
        const existing = uniqueTerms.get(normEn);
        // keep the one with more complete info (longer definition)
        const currentScore = (term.definitionEn?.length || 0) + (term.example?.length || 0);
        const existingScore = (existing.definitionEn?.length || 0) + (existing.example?.length || 0);
        
        if (currentScore > existingScore) {
            duplicates.push({ removed: existing.termEn, kept: term.termEn });
            uniqueTerms.set(normEn, term);
        } else {
            duplicates.push({ removed: term.termEn, kept: existing.termEn });
        }
    } else {
        uniqueTerms.set(normEn, term);
    }
}

// Rebuild the array and save
const newArrayStrs = [];
for (const [normEn, term] of uniqueTerms.entries()) {
    newArrayStrs.push(term._raw);
}

const beforeLen = terms.length;
const afterLen = newArrayStrs.length;

// Save summary report
const summary = {
    totalBefore: beforeLen,
    totalAfter: afterLen,
    duplicatesRemoved: duplicates.length,
    removedList: duplicates
};

fs.writeFileSync('scratch/dedup_summary.json', JSON.stringify(summary, null, 2));

// Replace the array in the original content
// Find the bounds of the array
const arrayStart = content.indexOf('export const terms: Term[] = [');
if (arrayStart !== -1) {
    const arrayBracketStart = content.indexOf('[', arrayStart);
    const arrayBracketEnd = content.lastIndexOf('];');
    
    if (arrayBracketStart !== -1 && arrayBracketEnd !== -1) {
        const before = content.substring(0, arrayBracketStart + 1);
        const after = content.substring(arrayBracketEnd);
        
        const newArrayContent = '\n  ' + newArrayStrs.join(',\n  ') + '\n';
        const newContent = before + newArrayContent + after;
        
        fs.writeFileSync(filePath, newContent);
        console.log(`Successfully written cleaned file. Saved ${duplicates.length} duplicates.`);
    } else {
        console.error("Could not find array brackets");
    }
} else {
    console.error("Could not find export statement");
}
